### PR TITLE
[PATCH v8] m4 macro usage without gpl part1

### PR DESCRIPTION
**Pull request by muvarov**

### CODE REVIEW

File: m4/odp_check_flag.m4 [hunk 1]
```diff
@@ -1,26 +1,40 @@
...
+	  AC_LANG_PUSH([CXX])
+	  ac_saved_cxxflags="$CXXFLAGS"
+	  CXXFLAGS="-W -Wall -Werror $ODP_CXXFLAGS $1"
+	  AC_COMPILE_IFELSE([AC_LANG_PROGRAM([])],
+		[AC_MSG_RESULT([yes])
+		ODP_CXXFLAGS="$ODP_CXXFLAGS $1"
+		CXXFLAGS="$ac_saved_cxxflags"],
+		[AC_MSG_RESULT([no])
+		CXXFLAGS="$ac_saved_cxxflags"])
+	  AC_LANG_POP([XXC])])
```
Comment (lumag) [inline m4/odp_check_flag.m4]: CXX

File: m4/odp_check_flag.m4 [hunk 2]
```diff
@@ -1,26 +1,40 @@
...
-	 [AC_LANG_PUSH([C++])
-	  AX_CHECK_COMPILE_FLAG([$1],
-				[ODP_CXXFLAGS="$ODP_CXXFLAGS $1"],
-				[], [-W -Wall -Werror],
-				[AC_LANG_SOURCE([int main(void)
-						{return 0;}])])
-	  AC_LANG_POP([C++])])
+[	  AC_MSG_CHECKING([if $CCXX supports $1])
+	  AC_LANG_PUSH([CXX])
+	  ac_saved_cxxflags="$CXXFLAGS"
```
Comment (lumag) [inline m4/odp_check_flag.m4]: We'd better not use `ac_` prefixed variables in our code.

File: m4/odp_check_flag.m4 [hunk 3]
```diff
@@ -1,26 +1,40 @@
...
-# Add FLAG to ODP_CFLAGS if compiler supports that option
 AC_DEFUN([ODP_CHECK_CXXFLAG],
-	 [AC_LANG_PUSH([C++])
-	  AX_CHECK_COMPILE_FLAG([$1],
-				[ODP_CXXFLAGS="$ODP_CXXFLAGS $1"],
-				[], [-W -Wall -Werror],
-				[AC_LANG_SOURCE([int main(void)
-						{return 0;}])])
-	  AC_LANG_POP([C++])])
+[	  AC_MSG_CHECKING([if $CCXX supports $1])
```
Comment (lumag) [inline m4/odp_check_flag.m4]: `$CXX`

File: m4/odp_check_flag.m4 [hunk 4]
```diff
@@ -1,26 +1,40 @@
...
+	  AC_LANG_PUSH([CXX])
```
Comment (lumag) [inline m4/odp_check_flag.m4]: C++

File: m4/odp_check_flag.m4 [hunk 5]
```diff
@@ -1,26 +1,40 @@
...
-	 [AC_LANG_PUSH([C])
-	  AX_CHECK_COMPILE_FLAG([$1],
-				[ODP_CFLAGS="$ODP_CFLAGS $1"],
-				[], [-W -Wall -Werror],
-				[AC_LANG_SOURCE([int main(void)
-						{return 0;}])])
+[
+	  AC_MSG_CHECKING([if $CC supports $1])
+	  AC_LANG_PUSH([C])
+	  ac_saved_cflags="$CFLAGS"
```
Comment (lumag) [inline m4/odp_check_flag.m4]: And here too.

File: m4/odp_check_flag.m4 [hunk 6]
```diff
@@ -1,26 +1,40 @@
...
+[	  AC_MSG_CHECKING([if $CCXX supports $1])
+	  AC_LANG_PUSH([CXX])
+	  ac_saved_cxxflags="$CXXFLAGS"
+	  CXXFLAGS="-W -Wall -Werror $ODP_CXXFLAGS $1"
+	  AC_COMPILE_IFELSE([AC_LANG_PROGRAM([])],
+		[AC_MSG_RESULT([yes])
+		ODP_CXXFLAGS="$ODP_CXXFLAGS $1"
+		CXXFLAGS="$ac_saved_cxxflags"],
+		[AC_MSG_RESULT([no])
+		CXXFLAGS="$ac_saved_cxxflags"])
```
Comment (lumag) [inline m4/odp_check_flag.m4]: Restoring of CFLAGS/CXXFLAGS can be moved outside of `AC_COMPILE_IFELSE`.

File: platform/linux-generic/m4/configure.m4 [hunk 7]
```diff
@@ -3,7 +3,10 @@ ODP_IMPLEMENTATION_NAME="odp-linux"
 ODP_VISIBILITY
 ODP_ATOMIC
 
-ODP_PTHREAD
+ODP_CHECK_CFLAG([-pthread])
+PTHREAD_LIBS_LIBODP="-lpthread"
+AC_SUBST([PTHREAD_LIBS_LIBODP])
```
Comment (lumag) [inline platform/linux-generic/m4/configure.m4]: 1) This can become implementation of `ODP_PTHREAD`, thus simplifying porting this to ODP-DPDK
2) Linker option is `-pthread`, not `-lpthread`, as far as I remember. And availability of that option (or library) also should be checked.

**Comment by lumag** [comment]
First patch looks good, other two patches should be fixed.

**Comment by muvarov** [comment]
@lumag  thanks for review. If general idea is on I will send v3.

**Comment by muvarov** [comment]
Interesting this PR now captured error on arm which is not related to current changes:

	cc:			clang-3.8 --target=arm-linux-gnueabihf
	cc version:             clang version 3.8.0-2ubuntu3~trusty5 (tags/RELEASE_380/final)

[1m../../platform/linux-generic/arch/arm/odp/api/abi/cpu.h:23:17: [0m[0;1;31merror: [0m[1m
      instruction requires: data-barriers[0m
        __asm volatile("isb" ::: "memory");
[0;1;32m                       ^
[0m[1m<inline asm>:1:2: [0m[0;1;30mnote: [0minstantiated into assembly here[0m
        isb
[0;1;32m        ^
[0m1 error generated.

**Comment by Bill-Fischofer-Linaro** [comment]
@muvarov Interesting. So is this a glitch or a real bug that's now been identified?

**Comment by muvarov** [comment]
this is how linux defines ISB
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/include/asm/barrier.h?h=v4.17
for arm > v7 it's isb::: memory, for v6 it's mcr p15 .. 
now I'm a little bit confused how it worked before...

**Comment by Bill-Fischofer-Linaro** [comment]
@brbrooks Can you advise on the Arm architecture aspects of this?

File: platform/linux-generic/arch/arm/odp/api/abi/cpu.h [hunk 8]
```diff
@@ -20,7 +20,8 @@ static inline void odp_cpu_pause(void)
 	 * ISB flushes the pipeline, then restarts. This is guaranteed to
 	 * stall the CPU a number of cycles.
 	 */
-	__asm volatile("isb" ::: "memory");
+	__asm volatile("isb sy" ::: "memory");
```
Comment (lumag) [inline platform/linux-generic/arch/arm/odp/api/abi/cpu.h]: commit description does not match code.

Comment (muvarov) [reply platform/linux-generic/arch/arm/odp/api/abi/cpu.h]: yea, I tried to use _buidin but this version of clang does not support it.

**Comment by muvarov** [comment]
@lumag this version (v7) is ready for  review. I hope I fixed all comments.

**Comment by brbrooks** [comment]
In AArch64/A64, "isb" and "isb sy" are the same. From clang, it seems "sy" is required for Thumb.

File: m4/odp_check_flag.m4 [hunk 9]
```diff
@@ -20,7 +20,7 @@ AC_DEFUN([ODP_CHECK_CFLAG],
 		[AC_MSG_RESULT([yes])
 		ODP_CFLAGS="$ODP_CFLAGS $1"],
 		[AC_MSG_RESULT([no])])
-	  CFLAGS="$ac_saved_cflags"
+	  CFLAGS="$saved_cflags"
```
Comment (lumag) [inline m4/odp_check_flag.m4]: This line does not belong to this patch

File: m4/odp_pthread.m4 [hunk 10]
```diff
@@ -1,7 +1,15 @@
...
 # -----------
 # Check for pthreads availability
-AC_DEFUN([ODP_PTHREAD], [dnl
-AX_PTHREAD([CC="$PTHREAD_CC"],
-	   [AC_MSG_FAILURE([We require pthreads to be available])])
+AC_DEFUN([ODP_PTHREAD], [
+	AC_MSG_CHECKING([for pthread support in -pthread])
+	AC_LANG_PUSH([C])
+	CFLAGS="$CFLAGS -pthread"
+	LDFLAGS="$LDFLAGS -pthread"
```
Comment (lumag) [inline m4/odp_pthread.m4]: I've mixed feelings towards this change. Why do you want to drop PTHREAD_CFLAGS/PTHREAD_LIBS? Several should have been compiled w/o pthreads, IIRC (proc mode ones).

File: include/odp/arch/arm32-linux/odp/api/abi/cpu.h [hunk 11]
```diff
@@ -20,7 +20,11 @@ static inline void odp_cpu_pause(void)
 	 * ISB flushes the pipeline, then restarts. This is guaranteed to
 	 * stall the CPU a number of cycles.
 	 */
-	__asm volatile("isb" ::: "memory");
+#if __ARM_ARCH < 7
+	__asm__ __volatile__("isb" : : : "memory");
+#else
+	__asm__ __volatile__("isb sy" : : : "memory");
+#endif
```
[the remark this answers is not in the document]
Comment (muvarov) [reply include/odp/arch/arm32-linux/odp/api/abi/cpu.h]: I saw error with clang compilation while gcc was fine. I can recheck if this patch now optional or not.

[the remark this answers is not in the document]
Comment (lumag) [reply include/odp/arch/arm32-linux/odp/api/abi/cpu.h]: @muvarov Well, I'm just not sure how does changing M4 macros affect compilation. Could you please recheck? Because if you saw this error (btw, could you please include it into commit message) after changing macros, it can be a sign of some hidden behavioural change.

**Comment by muvarov** [comment]
@lumag CI was ok without isb patch. I will track this issue separately. I think it has to be shown if --target is specified but --march=armv7-a was not specified. Or something like that.

In v8 change is in pthread patch.

**Comment by lumag** [comment]
Reviewed-by: Dmitry Eremin-Solenikov

**Comment by muvarov** [comment]
Merged.
